### PR TITLE
chore: make message logs maintainable - WPB-9517

### PR DESCRIPTION
**Pull request by markusfassbender**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-9517" title="WPB-9517" target="_blank">BugWPB-9517</a>  [iOS] Assets frequently fail to send
  </td></table>
  <br />
 

### Issue

There are two protocols `ProteusMessage` and `MLSMessage`, that are combined:
`public typealias SendableMessage = ProteusMessage & MLSMessage`.

The old implementation made heavy use of implicit Swift behaviour and was difficult to understand:
- `ProteusMessage` implemented logging for `ZMClientMessage`, but `MLSMessage` did not
- `MLSMessage` implemented logging for `ZMAssetClientMessage`, but  `ProteusMessage` did not

Now there is a non-protocol extension solution (a simple struct) `MessageLogAttributes` that just take types as function parameters. For convenience it still needed to support `SendableMessage` and `ProteusMessage`, but adds an assertion to detect uncovered cases.

The struct also ensure a `context.perform` to access the properties correctly and avoid crashes.

### Testing

- Regression test all kind messages and look into the logs.

---

### Checklist

- [x] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [x] Description is filled and free of optional paragraphs.
- [x] Adds/updates automated tests.

### CODE REVIEW

File: wire-ios-request-strategy/Sources/Request Strategies/Client Message/ClientMessageRequestStrategy.swift [hunk 1]
```diff
@@ -90,22 +90,31 @@ extension ClientMessageRequestStrategy: InsertedObjectSyncTranscoder {
     typealias Object = ZMClientMessage
 
     func insert(object: ZMClientMessage, completion: @escaping () -> Void) {
-        WireLogger.messaging.debug("inserting message", attributes: object.logInformation)
+        let messageLogAttributes = MessageLogAttributes(context: context)
+
+        Task {
+            let logInformation = await messageLogAttributes.logAttributes(object)
+            WireLogger.messaging.debug("inserting message", attributes: logInformation)
+        }
```
Comment (markusfassbender) [inline wire-ios-request-strategy/Sources/Request Strategies/Client Message/ClientMessageRequestStrategy.swift]: Before some logs like here access `.logInformation` without `context.perform`. Even though it works for those cases, copying it to other places might crash the app.
Now it always ensures to access properties via `context.perform`.

File: wire-ios-request-strategy/Sources/Object Syncs/Helpers/ProteusMessage.swift [hunk 2]
```diff
@@ -18,21 +18,8 @@
 
 import Foundation
 
-public protocol ProteusMessage: OTREntity, EncryptedPayloadGenerator, Hashable {
-    var logInformation: LogAttributes { get }
-}
+public protocol ProteusMessage: OTREntity, EncryptedPayloadGenerator {}
```
Comment (markusfassbender) [inline wire-ios-request-strategy/Sources/Object Syncs/Helpers/ProteusMessage.swift:21]: Maybe the whole protocol is now obsolete, but still used in multiple places I didn't want to refactor with this PR.

File: wire-ios-request-strategy/Sources/Helpers/MessageExpirationTimer.swift [hunk 3]
```diff
@@ -79,20 +79,25 @@ public class MessageExpirationTimer: ZMMessageTimer, ZMContextChangeTracker {
...
             }
         }
     }
 
+    private func logDebug(_ text: String, message: any ProteusMessage) {
+        Task {
+            let logInformation = await MessageLogAttributes(context: moc).logAttributes(message)
+            WireLogger.messaging.debug(text, attributes: logInformation)
+        }
+    }
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Helpers/MessageExpirationTimer.swift]: why in a Task ? do we have a guarantee that first log is finished before second one here?

File: wire-ios-request-strategy/Sources/Request Strategies/Client Message/ClientMessageRequestStrategy.swift (hunk 1 above)
Comment (netbe) [reply wire-ios-request-strategy/Sources/Request Strategies/Client Message/ClientMessageRequestStrategy.swift]: same here with the task I am not sure we can guarantee it's done before next task starts.

I would go for non async version of it with context.performAndWait as it allows reentrancy (all strategy methods are supposed to be called from syncContext here)

File: wire-ios-request-strategy/Sources/Object Syncs/Helpers/MessageLogAttributes.swift [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        if let assetClientMessage = message as? ZMAssetClientMessage {
+            return await logAttributes(assetClientMessage)
+        }
+
+        if let genericMessage = message as? GenericMessageEntity {
+            return await logAttributes(genericMessage)
+        }
+
+        assertionFailure("cannot find a supported type of message '\(type(of: message))'")
+        return [:]
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Object Syncs/Helpers/MessageLogAttributes.swift]: nice :)

File: wire-ios-request-strategy/Sources/Object Syncs/Helpers/MessageLogAttributes.swift [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import CoreData
+import WireDataModel
+
+/// Provides log attributes for messages of supported message types.
+struct MessageLogAttributes {
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Object Syncs/Helpers/MessageLogAttributes.swift]: suggestion: MessageLogAttributesBuilder

**Review by netbe — COMMENTED** [review]
Added some concerns about Tasks

File: wire-ios-request-strategy/Sources/Request Strategies/Client Message/ClientMessageRequestStrategy.swift (hunk 1 above)
Comment (markusfassbender) [reply wire-ios-request-strategy/Sources/Request Strategies/Client Message/ClientMessageRequestStrategy.swift]: Done this way, thanks

File: wire-ios-request-strategy/Sources/Helpers/MessageExpirationTimer.swift (hunk 3 above)
Comment (markusfassbender) [reply wire-ios-request-strategy/Sources/Helpers/MessageExpirationTimer.swift]: made sync with `context.performAndWait`

**Review by netbe — APPROVED** [review]

**Review by caldrian — APPROVED** [review]

**Pull request merged**